File: backend/internal/provider/extensions/netease.go
package extensions

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"

	"github.com/guohuiyuan/go-music-dl/internal/provider/model"
)

const neteaseBaseURL = "https://music.163.com"

type Netease struct {
	Cookie string
}

func NewNetease(cookie string) *Netease {
	return &Netease{Cookie: strings.TrimSpace(cookie)}
}

func (client *Netease) SearchPlaylist(keyword string) ([]model.Playlist, error) {
	payload, err := client.search(keyword, 1000)
	if err != nil {
		return nil, err
	}
	return neteasePlaylists(at(payload, "result", "playlists")), nil
}

func (client *Netease) SearchAlbum(keyword string) ([]model.Playlist, error) {
	payload, err := client.search(keyword, 10)
	if err != nil {
		return nil, err
	}
	return neteaseAlbums(at(payload, "result", "albums")), nil
}

func (client *Netease) search(keyword string, searchType int) (map[string]interface{}, error) {
	form := url.Values{
		"s": {strings.TrimSpace(keyword)}, "type": {strconv.Itoa(searchType)},
		"limit": {"30"}, "offset": {"0"},
	}
	return postFormJSON(neteaseBaseURL+"/api/search/get/web", client.Cookie, form)
}

func (client *Netease) GetPlaylistSongs(id string) ([]model.Song, error) {
	_, songs, err := client.Playlist(id)
	return songs, err
}

func (client *Netease) GetAlbumSongs(id string) ([]model.Song, error) {
	_, songs, err := client.Album(id)
	return songs, err
}

func (client *Netease) Playlist(id string) (*model.Playlist, []model.Song, error) {
	endpoint := neteaseBaseURL + "/api/v6/playlist/detail?id=" + url.QueryEscape(strings.TrimSpace(id)) + "&n=1000&s=8"
	payload, err := getJSON(endpoint, client.Cookie, nil)
	if err != nil {
		return nil, nil, err
	}
	rawPlaylist := object(payload["playlist"])
	if len(rawPlaylist) == 0 {
		return nil, nil, errorsForEmpty("netease playlist")
	}
	playlist := neteasePlaylist(rawPlaylist)
	return &playlist, neteaseSongs(rawPlaylist["tracks"]), nil
}

func (client *Netease) ParsePlaylist(link string) (*model.Playlist, []model.Song, error) {
	id := linkID(link, "id")
	if id == "" {
		return nil, nil, fmt.Errorf("netease playlist link has no id")
	}
	return client.Playlist(id)
}

func (client *Netease) Album(id string) (*model.Playlist, []model.Song, error) {
	payload, err := getJSON(neteaseBaseURL+"/api/album/"+url.PathEscape(strings.TrimSpace(id)), client.Cookie, nil)
	if err != nil {
		return nil, nil, err
	}
	rawAlbum := object(payload["album"])
	if len(rawAlbum) == 0 {
		return nil, nil, errorsForEmpty("netease album")
	}
	album := neteaseAlbum(rawAlbum)
	return &album, neteaseSongs(payload["songs"]), nil
}

func (client *Netease) ParseAlbum(link string) (*model.Playlist, []model.Song, error) {
	id := linkID(link, "id")
	if id == "" {
		return nil, nil, fmt.Errorf("netease album link has no id")
	}
	return client.Album(id)
}

func (client *Netease) RecommendedPlaylists() ([]model.Playlist, error) {
	payload, err := getJSON(neteaseBaseURL+"/api/personalized/playlist?limit=30", client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	return neteasePlaylists(payload["result"]), nil
}

func (client *Netease) PlaylistCategories() ([]model.PlaylistCategory, error) {
	values := []struct{ group, name string }{
		{"语种", "华语"}, {"语种", "欧美"}, {"语种", "日语"}, {"语种", "韩语"},
		{"风格", "流行"}, {"风格", "摇滚"}, {"风格", "民谣"}, {"风格", "电子"},
		{"场景", "学习"}, {"场景", "运动"}, {"场景", "夜晚"}, {"情感", "怀旧"},
	}
	result := make([]model.PlaylistCategory, 0, len(values))
	for _, value := range values {
		result = append(result, model.PlaylistCategory{ID: value.name, Name: value.name, Group: value.group, Source: "netease"})
	}
	return result, nil
}

func (client *Netease) CategoryPlaylists(categoryID string, page, limit int) ([]model.Playlist, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 30
	}
	query := url.Values{
		"cat": {strings.TrimSpace(categoryID)}, "order": {"hot"}, "total": {"true"},
		"limit": {strconv.Itoa(limit)}, "offset": {strconv.Itoa((page - 1) * limit)},
	}
	payload, err := getJSON(neteaseBaseURL+"/api/playlist/list?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	return neteasePlaylists(payload["playlists"]), nil
}

func (client *Netease) UserPlaylists(page, limit int) ([]model.Playlist, error) {
	account, err := getJSON(neteaseBaseURL+"/api/nuser/account/get", client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	userID := text(at(account, "profile", "userId"))
	if userID == "" {
		userID = text(at(account, "account", "id"))
	}
	if userID == "" {
		return nil, fmt.Errorf("netease login cookie is required")
	}
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 50
	}
	query := url.Values{"uid": {userID}, "limit": {strconv.Itoa(limit)}, "offset": {strconv.Itoa((page - 1) * limit)}}
	payload, err := getJSON(neteaseBaseURL+"/api/user/playlist?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	return neteasePlaylists(payload["playlist"]), nil
}

func (client *Netease) IsVIPAccount() (bool, error) {
	payload, err := getJSON(neteaseBaseURL+"/api/nuser/account/get", client.Cookie, nil)
	if err != nil {
		return false, err
	}
	if text(at(payload, "account", "id")) == "" {
		return false, fmt.Errorf("netease login cookie is invalid")
	}
	return integer(at(payload, "profile", "vipType")) > 0 || integer(at(payload, "account", "vipType")) > 0, nil
}

func neteaseSongs(value interface{}) []model.Song {
	items := array(value)
	result := make([]model.Song, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := text(raw["id"])
		if id == "" {
			continue
		}
		album := object(raw["al"])
		if len(album) == 0 {
			album = object(raw["album"])
		}
		artists := raw["ar"]
		if artists == nil {
			artists = raw["artists"]
		}
		duration := integer(raw["dt"])
		if duration == 0 {
			duration = integer(raw["duration"])
		}
		result = append(result, model.Song{
			ID: id, Name: text(raw["name"]), Artist: joinNames(artists, "name"),
			Album: text(album["name"]), AlbumID: text(album["id"]), Duration: duration / 1000,
			Source: "netease", Cover: text(album["picUrl"]),
			Extra: map[string]string{"song_id": id, "album_id": text(album["id"])},
		})
	}
	return result
}

func neteasePlaylists(value interface{}) []model.Playlist {
	items := array(value)
	result := make([]model.Playlist, 0, len(items))
	for _, item := range items {
		playlist := neteasePlaylist(object(item))
		if playlist.ID != "" {
			result = append(result, playlist)
		}
	}
	return result
}

func neteasePlaylist(raw map[string]interface{}) model.Playlist {
	cover := text(raw["coverImgUrl"])
	if cover == "" {
		cover = text(raw["picUrl"])
	}
	return model.Playlist{
		ID: text(raw["id"]), Name: text(raw["name"]), Cover: cover,
		TrackCount: integer(raw["trackCount"]), PlayCount: integer(raw["playCount"]),
		Creator: text(at(raw, "creator", "nickname")), Description: text(raw["description"]),
		Source: "netease", Link: "https://music.163.com/playlist?id=" + text(raw["id"]),
	}
}

func neteaseAlbums(value interface{}) []model.Playlist {
	items := array(value)
	result := make([]model.Playlist, 0, len(items))
	for _, item := range items {
		album := neteaseAlbum(object(item))
		if album.ID != "" {
			result = append(result, album)
		}
	}
	return result
}

func neteaseAlbum(raw map[string]interface{}) model.Playlist {
	return model.Playlist{
		ID: text(raw["id"]), Name: text(raw["name"]), Cover: text(raw["picUrl"]),
		TrackCount: integer(raw["size"]), Creator: joinNames(raw["artists"], "name"),
		Description: text(raw["description"]), Source: "netease",
		Link: "https://music.163.com/album?id=" + text(raw["id"]),
	}
}

func errorsForEmpty(kind string) error {
	return fmt.Errorf("%s response is empty", kind)
}

File: backend/provider_bridge/platforms/netease.py
"""NetEase collection adapter hosted by the provider sidecar."""

from __future__ import annotations

from urllib.parse import quote, urlencode

from provider_bridge.collection_common import (
    at,
    category_payload,
    collection_payload,
    integer,
    join_names,
    link_id,
    list_value,
    object_value,
    page_limit,
    string,
    track_payload,
)
from provider_bridge.platform_http import PlatformHTTP


BASE_URL = "https://music.163.com"


class NeteaseCollections:
    source = "netease"

    def __init__(self, cookie: str = "", session=None):
        self.http = PlatformHTTP(cookie, session)

    def search_playlist(self, keyword: str, **_):
        return self._playlists(at(self._search(keyword, 1000), "result", "playlists"))

    def search_album(self, keyword: str, **_):
        return self._albums(at(self._search(keyword, 10), "result", "albums"))

    def _search(self, keyword: str, search_type: int):
        return self.http.post_form(
            BASE_URL + "/api/search/get/web",
            {"s": string(keyword), "type": search_type, "limit": 30, "offset": 0},
        )

    def playlist(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("netease playlist id is empty")
        payload = self.http.get(
            BASE_URL + "/api/v6/playlist/detail?" + urlencode({"id": identifier, "n": 1000, "s": 8})
        )
        raw = object_value(payload.get("playlist"))
        if not raw:
            raise ValueError("netease playlist response is empty")
        return self._playlist(raw), self._songs(raw.get("tracks"))

    def album(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("netease album id is empty")
        payload = self.http.get(BASE_URL + "/api/album/" + quote(identifier, safe=""))
        raw = object_value(payload.get("album"))
        if not raw:
            raise ValueError("netease album response is empty")
        return self._album(raw), self._songs(payload.get("songs"))

    def parse_playlist(self, link: str, **_):
        identifier = link_id(link, "id")
        if not identifier:
            raise ValueError("netease playlist link has no id")
        return self.playlist(identifier)

    def parse_album(self, link: str, **_):
        identifier = link_id(link, "id")
        if not identifier:
            raise ValueError("netease album link has no id")
        return self.album(identifier)

    def recommend(self, **_):
        payload = self.http.get(BASE_URL + "/api/personalized/playlist?limit=30")
        return self._playlists(payload.get("result"))

    def categories(self, **_):
        values = (
            ("语种", "华语"), ("语种", "欧美"), ("语种", "日语"), ("语种", "韩语"),
            ("风格", "流行"), ("风格", "摇滚"), ("风格", "民谣"), ("风格", "电子"),
            ("场景", "学习"), ("场景", "运动"), ("场景", "夜晚"), ("情感", "怀旧"),
        )
        return [category_payload(identifier=name, source=self.source, name=name, group=group) for group, name in values]

    def category(self, category_id: str, page: int = 1, limit: int = 30, **_):
        page, limit = page_limit(page, limit)
        query = urlencode({
            "cat": string(category_id), "order": "hot", "total": "true",
            "limit": limit, "offset": (page - 1) * limit,
        })
        return self._playlists(self.http.get(BASE_URL + "/api/playlist/list?" + query).get("playlists"))

    def user_playlists(self, page: int = 1, limit: int = 50, **_):
        account = self.http.get(BASE_URL + "/api/nuser/account/get")
        user_id = string(at(account, "profile", "userId")) or string(at(account, "account", "id"))
        if not user_id:
            raise ValueError("netease login cookie is required")
        page, limit = page_limit(page, limit, 50)
        query = urlencode({"uid": user_id, "limit": limit, "offset": (page - 1) * limit})
        return self._playlists(self.http.get(BASE_URL + "/api/user/playlist?" + query).get("playlist"))

    def _songs(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = string(raw.get("id"))
            if not identifier:
                continue
            album = object_value(raw.get("al")) or object_value(raw.get("album"))
            artists = raw.get("ar") if raw.get("ar") is not None else raw.get("artists")
            duration_ms = integer(raw.get("dt")) or integer(raw.get("duration"))
            album_id = string(album.get("id"))
            result.append(track_payload(
                identifier=identifier,
                source=self.source,
                name=string(raw.get("name")),
                artist=join_names(artists),
                album=string(album.get("name")),
                album_id=album_id,
                duration=duration_ms // 1000,
                cover=string(album.get("picUrl")),
                extra={"song_id": identifier, "album_id": album_id},
            ))
        return result

    def _playlists(self, value):
        return [item for item in (self._playlist(object_value(raw)) for raw in list_value(value)) if item["id"]]

    def _playlist(self, raw):
        identifier = string(raw.get("id"))
        cover = string(raw.get("coverImgUrl")) or string(raw.get("picUrl"))
        return collection_payload(
            identifier=identifier,
            source=self.source,
            name=string(raw.get("name")),
            creator=string(at(raw, "creator", "nickname")),
            description=string(raw.get("description")),
            cover=cover,
            link="https://music.163.com/playlist?id=" + identifier,
            track_count=integer(raw.get("trackCount")),
            play_count=integer(raw.get("playCount")),
        )

    def _albums(self, value):
        return [item for item in (self._album(object_value(raw)) for raw in list_value(value)) if item["id"]]

    def _album(self, raw):
        identifier = string(raw.get("id"))
        return collection_payload(
            identifier=identifier,
            source=self.source,
            name=string(raw.get("name")),
            creator=join_names(raw.get("artists")),
            description=string(raw.get("description")),
            cover=string(raw.get("picUrl")),
            link="https://music.163.com/album?id=" + identifier,
            track_count=integer(raw.get("size")),
        )

